### ISSUE TITLE
Loss values

### ISSUE THREAD

**Issue by mshooter** [anibali/margipose#37]
Could you tell me if your values are getting negative? 
When I use the loss described in your paper I am getting negative values

**Comment by anibali** [comment]
That should not be possible since each term in equation (6) of the paper is non-negative (l2 loss can't be negative, and [JSD is lower-bounded by zero](https://en.wikipedia.org/wiki/Jensen%E2%80%93Shannon_divergence#Bounds)). Make sure that you have a) normalised the heatmaps, and b) implemented JSD correctly. My guess is that the heatmaps have not been normalised (each 2D heatmap should sum to one).

I assume that you are trying to create your own implementation of a MargiPose-like loss function without using this package. These days I suggest using [Kornia](https://kornia.readthedocs.io/en/latest/) to assist with this, which provides:

- [spatial_softmax2d](https://kornia.readthedocs.io/en/latest/geometry.subpix.html#kornia.geometry.subpix.spatial_softmax2d) - normalises heatmaps, i.e. the last block in Figure 6. Note that you **can't just use torch.nn.Softmax2d** since that function operates along a different tensor axis.
- [spatial_expectation2d](https://kornia.readthedocs.io/en/latest/geometry.subpix.html#kornia.geometry.subpix.spatial_expectation2d) - calculates the expectation of normalised heatmaps. This function can be used to help calculate mu, see equation (3). For more details on how to do this, look at [heatmaps_to_coords](https://github.com/anibali/margipose/blob/604494922989b58a5f1fa5d2ecfbe7e062578dae/src/margipose/models/margipose_model.py#L254-L261) but replace `dsnt` with `spatial_expectation2d`.
- [js_div_loss_2d](https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.js_div_loss_2d) - calculates JSD between two normalised heatmaps.

**Comment by mshooter** [comment]
Thank you for your suggestions! 
I did try not to implement it myself I mostly used the code that was available.
However, I did have to change certain things depending on the dataset.
